### ISSUE TITLE
support for batch norm with 1D CNN seems to be missing

### ISSUE THREAD

**Issue by MPdaedalus** [deeplearning4j/deeplearning4j#9263]
batch normalization for 1DConv does not work due to 1DConv requiring rank 3 INDArrays and BatchNormalization.java only supporting rank 2 or 4 INDArrays. 

Example layer config>
.addLayer("stem-cnn1",new Convolution1DLayer.Builder(3,2).nIn(483).nOut(32).build(),"input")
.addLayer("stem-batch1", new BatchNormalization.Builder(false).decay(0.995).eps(0.001).nIn(32).nOut(32).build(),"stem-cnn1")

 input created from 2D float array and converted to rank3 via TimeSeriesUtils
allData = new DataSet();
allData.setFeatures(TimeSeriesUtils.reshape2dTo3d(Nd4j.create(features.toArray(new float[features.size()][])),64));
allData.setLabels(TimeSeriesUtils.reshape2dTo3d(Nd4j.create(labels.toArray(new float[labels.size()][])),64));

when I run model.fit(allData)   >

Exception in thread "main" java.lang.IllegalStateException: Batch normalization on activations of rank 3 not supported (layer name: stem-batch1, layer index: 0, layer type: BatchNormalization)
	at org.deeplearning4j.nn.layers.normalization.BatchNormalization.preOutput(BatchNormalization.java:506)
	at org.deeplearning4j.nn.layers.normalization.BatchNormalization.activate(BatchNormalization.java:393)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(LayerVertex.java:110)
	at org.deeplearning4j.nn.graph.ComputationGraph.ffToLayerActivationsInWS(ComputationGraph.java:2134)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1371)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1340)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:174)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:61)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1164)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1114)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1081)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1017)
	at main.PatternDetect.run(PatternDetect.java:92)
	at main.Start.main(Start.java:15)

this occurs after cuDNN failure (only 4D and 5D tensors supported in cuDNN)

 WARN org.deeplearning4j.nn.layers.normalization.BatchNormalization - CuDNN BatchNormalization forward pass execution failed - falling back on built-in implementation
java.lang.IllegalArgumentException: Invalid size: cannot get size of dimension 3 for rank 3 NDArray (array shape: [57, 32, 246])
	at org.nd4j.linalg.api.ndarray.BaseNDArray.size(BaseNDArray.java:4529)
	at org.deeplearning4j.cuda.normalization.CudnnBatchNormalizationHelper.preOutput(CudnnBatchNormalizationHelper.java:265)
	at org.deeplearning4j.nn.layers.normalization.BatchNormalization.preOutput(BatchNormalization.java:451)
	at org.deeplearning4j.nn.layers.normalization.BatchNormalization.activate(BatchNormalization.java:393)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(LayerVertex.java:110)
	at org.deeplearning4j.nn.graph.ComputationGraph.ffToLayerActivationsInWS(ComputationGraph.java:2134)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1371)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1340)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:174)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:61)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1164)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1114)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1081)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1017)

offending code in BatchNormalization.java

NDArray mean, var;
        if (training == TrainingMode.TRAIN) {
            switch (x.rank()) {
                case 2:
                    // mean and variance over samples in batch
                    mean = x.mean(0);
                    var = x.var(false, 0);
                    break;
                case 4:
                    // mean and variance over samples AND locations

                    mean = x.mean(nonChDims);
                    var = x.var(false, nonChDims);
                    break;
                default:
                    throw new IllegalStateException("Batch normalization on activations of rank " + x.rank()
                            + " not supported " + layerId());
            }

should it not be possible to add a "case 3" and do mean and variance over the correct dimension?
As it stands building deep 1DConv nets is impossible in DL4J because BN is required to prevent exploding gradients.

on DL4J Snapshot cuda 11.2 with cuDNN on  fedora 32 Kernel 5.11.11

other people have also run into this bug last year it seems>
https://community.konduit.ai/t/batchnormalization-layer-only-support-single-int-as-nin-and-nout/900

**Comment by agibsonccc** [comment]
@MPdaedalus if you want, please feel free to take a crack at the fix. Instead of building everything from source, you can configure snapshots so you don't have to build from source for the tests. Otherwise, I'll take a crack at this after release.

**Comment by MPdaedalus** [comment]
its a bit beyond what i'm used to (just learning to run and test CNNs) but I'll have a go, I don't really know the internals of DL4J beyond the basics so I might ask for some pointers here. 

How do I configure snapshots so you don't have to build from source for the tests and which tests should i run?

**Comment by agibsonccc** [comment]
@mpazos See: https://deeplearning4j.konduit.ai/config/config-snapshots thanks for being open to this!

**Comment by MPdaedalus** [comment]
I've tried to implement a solution as much as I think it is correct but no luck due to a bunch of errors downstream which is beyond my understanding. It was only a few lines of code that I changed which I will outline below. I decided to only test using CPU build instead of worrying about cuda. However even if the CPU version of batch norm worked I fear the performance would be really bad as batch norm is usually a heavy operation so it would probably have required a cuDNN implementation to be useful anyway which would require massaging it into a 4D tensor somehow.  For now I am going to try to use the standard ConvolutionLayer and just use 1x3,1x7 filters instead of 3x3,7x7, hopefully I should be able to do that with cuDNN batchnorm until you get around to implementing a proper 1D cuDNN batchnorm in the future. Hopefully the code below should provide you with a starting point to implement batch norm for cpu at least.

changes from line 492 to 514 of org.deeplearning4j.nn.layers.normalization.BatchNormalization.java

The idea being to add case for rank 3 and do mean and variance over the non channel dimensions which for 1DCNN with format RNNFormat.NCW  should be dimensions 0 and 2 hence>tdshape =new int[] {0,2};

 int[] tdshape;
        if (training == TrainingMode.TRAIN) {
            switch (x.rank()) {
                case 2:
                    // mean and variance over samples in batch
                    mean = x.mean(0);
                    var = x.var(false, 0);
                    break;
                case 3:
	                tdshape =new int[] {0,2};
	                mean = x.mean(tdshape);
	                var = x.var(false,tdshape);
	                break;
                case 4:
                    // mean and variance over samples AND locations
                    mean = x.mean(nonChDims);
                    var = x.var(false, nonChDims);
                    break;
                default:
                    throw new IllegalStateException("Batch normalization on activations of rank " + x.rank()
                            + " not supported " + layerId());
            }

plus change the second part of the next if/else block from
 } else if (x.rank() == 4) {
to  
} else if (x.rank() == 3 || x.rank() == 4) {

I end up with the following downstream error>

`Exception in thread "main" java.lang.RuntimeException: Op [maxpool2d] execution failed
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1583)
	at org.nd4j.linalg.factory.Nd4j.exec(Nd4j.java:6566)
	at org.deeplearning4j.nn.layers.convolution.subsampling.SubsamplingLayer.activate(SubsamplingLayer.java:305)
	at org.deeplearning4j.nn.layers.convolution.subsampling.Subsampling1DLayer.activate(Subsampling1DLayer.java:85)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(LayerVertex.java:110)
	at org.deeplearning4j.nn.graph.ComputationGraph.ffToLayerActivationsInWS(ComputationGraph.java:2134)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1371)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1340)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:174)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:61)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1164)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1114)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1081)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1017)
	at main.PatternDetect.run(PatternDetect.java:92)
	at main.Start.main(Start.java:15)
Caused by: java.lang.RuntimeException: Output array did not match expected shape.
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1918)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1562)
	... 16 more`

and also 

`java.lang.RuntimeException: Output array did not match expected shape.
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1918)
	at org.nd4j.linalg.factory.Nd4j.exec(Nd4j.java:6575)
	at org.deeplearning4j.nn.layers.mkldnn.MKLDNNSubsamplingHelper.activate(MKLDNNSubsamplingHelper.java:158)
	at org.deeplearning4j.nn.layers.convolution.subsampling.SubsamplingLayer.activate(SubsamplingLayer.java:251)
	at org.deeplearning4j.nn.layers.convolution.subsampling.Subsampling1DLayer.activate(Subsampling1DLayer.java:85)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(LayerVertex.java:110)
	at org.deeplearning4j.nn.graph.ComputationGraph.ffToLayerActivationsInWS(ComputationGraph.java:2134)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1371)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:1340)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:174)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:61)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.graph.ComputationGraph.fitHelper(ComputationGraph.java:1164)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1114)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1081)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:1017)
	at main.PatternDetect.run(PatternDetect.java:92)
	at main.Start.main(Start.java:15)
Expected vs provided shapes mismatch [57, 64, 122, 1] vs [57, 64, 123, 1] at index 0 with expected shape info [4, 57, 64, 122, 1, 7808, 122, 1, 1, 8192, 1, 99] and output shape info [4, 57, 64, 123, 1, 7872, 123, 1, 1, 8192, 1, 99]. Conditions, shapeEquals: 0, array empty: 0
19:52:33.296 [main] ERROR org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner - Failed to execute op maxpool2d. Attempted to execute with 1 inputs, 1 outputs, 0 targs,0 bargs and 11 iargs. Inputs: [(FLOAT,[57,64,246,1],c)]. Outputs: [(FLOAT,[57,64,123,1],c)]. tArgs: -. iArgs: [3, 1, 2, 2, 0, 0, 1, 1, 0, 0, 0]. bArgs: -. Op own name: "fce50f8f-0aca-4465-8c46-1f549a540280" - Please see above message (printed out from c++) for a possible cause of error.`

The network layers are the same as my first post with 1 1dcnn layer and a batch norm layer.

**Comment by agibsonccc** [comment]
@MPdaedalus  JFYI https://github.com/eclipse/deeplearning4j/pull/9338
